rust/src/memcap.rs

```diff
@@ -0,0 +1,257 @@
+/* Copyright (C) 2021 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+const BASE: isize = 2;
+
+/// enum to store currently allowed memory units in config,
+/// can be easily expanded for more units
+#[derive(Debug, Eq, PartialEq)]
+pub enum MemUnit {
+    B       = BASE.pow(0),
+    KB      = BASE.pow(10),
+    MB      = BASE.pow(20),
+    GB      = BASE.pow(30),
+    Unknown = 0,
+ }
+
+/// Helper function to retrieve memory unit from a string slice
+///
+/// Return value:
+///     MemUnit enum type
+///
+/// # Arguments
+///
+/// * `unit` - A string slice possibly containg memory unit
+///
+/// # Examples
+/// ```
+/// use suricata_rust::memcap::{self, MemUnit};
+///
+/// let unit = "KB";
+/// assert_eq!(MemUnit::KB, memcap::get_memunit(unit));
+///
+/// let unit = "mb";
+/// assert_eq!(MemUnit::MB, memcap::get_memunit(unit));
+///
+/// let unit = "gB";
+/// assert_eq!(MemUnit::GB, memcap::get_memunit(unit));
+/// ```
+pub fn get_memunit(unit: &str) -> MemUnit {
+    let unit = &unit.to_lowercase()[..];
+    match unit {
+        "b"     => { MemUnit::B }
+        "kb"    => { MemUnit::KB }
+        "mb"    => { MemUnit::MB }
+        "gb"    => { MemUnit::GB }
+        _       => { MemUnit::Unknown }
+    }
+}
+
+/// Parses memory units from human readable form to machine readable
+///
+/// Return value:
+///     Result => Ok(f64)
+///            => Err(error string)
+///
+/// # Arguments
+///
+/// * `arg` - A string slice that holds the value parsed from the config
+///
+/// # Examples
+///
+/// ```
+/// use suricata_rust::memcap::{self, MemUnit};
+///
+/// let s = "5kb";
+/// let res = (5 * 1024) as f64;
+/// assert_eq!(Ok(res), memcap::parse(s));
+///
+/// let s = "10";
+/// let res = 10 as f64;
+/// assert_eq!(Ok(res), memcap::parse(s));
+///
+/// let s = "4hb";
+/// assert_eq!(true, memcap::parse(s).is_err());
+/// ```
+pub fn parse(arg: &str) -> Result<f64, &'static str> {
```

Comment:
I see that the C variation uses double, but I'm not sure why.  I think we should return u64 though.

Reply:
Should it then be `ceil(any float memcap)`?

Reply:
Maybe I don't understand, but we also have "<integer> <unit>" right.. With unit defaulting to "bytes" if not set. Oh, we support "12.5 gb", etc..  I'd use `floor` to get the same behaviour we have in C.

I general I think all consumers of this would expect an integer type, not a float.